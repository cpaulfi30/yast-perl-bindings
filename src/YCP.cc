/**
 * $Id$
 *
 * This is the path from Perl to YCP. It defines XSUBs.
 */

#include <y2/Y2Namespace.h>
#include <y2/Y2Component.h>
#include <y2/Y2ComponentCreator.h>
#include <ycp/y2log.h>
#include <ycp/YBlock.h>
#include <ycp/YExpression.h>
#include <ycp/YStatement.h>

#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include "YPerl.h"

static
Y2Namespace *
getNs (const char * ns_name, const char * func_name)
{
    // func_name is only for debugging
    Y2Component *c = Y2ComponentBroker::getNamespaceComponent (ns_name);
    if (c == NULL)
    {
	y2error ("No component can provide namespace %s for a Perl call of %s",
		 ns_name, func_name);
	return NULL;
    }

    Y2Namespace *ns = c->import (ns_name);
    if (ns == NULL)
    {
	y2error ("Component %p (%s) could not provide namespace %s for a Perl call of %s",
		 c, c->name ().c_str (), ns_name, func_name);
    }
    else
    {
	ns->initialize ();
    }
    return ns;
}

// forward declaration
YCPValue YCP_call_SCR (pTHX_ const char * func_name, const vector<SV *>& args);
YCPValue YCP_getset_variable (pTHX_ const char * ns_name, SymbolEntryPtr var_se, const vector<SV *>& args);

XS(XS_YCP_y2_logger); /* prototype to pass -Wmissing-prototypes */
XS(XS_YCP_y2_logger)
{
    // defines "items", the number of arguments
    dXSARGS;
    if (items != 6)
    {
	y2internal ("y2_logger must have 6 arguments");
	XSRETURN_EMPTY;
    }
    loglevel_t level = (loglevel_t) SvIV (ST (0));
    const char * comp = SvPV_nolen (ST (1));
    const char * file = SvPV_nolen (ST (2));
    int line = SvIV (ST (3));
    const char * function = SvPV_nolen (ST (4));
    const char * message = SvPV_nolen (ST (5));
    y2_logger (level, comp, file, line, function, "%s", message);
    XSRETURN_EMPTY;
}

XS(XS_YCP_call_ycp); /* prototype to pass -Wmissing-prototypes */
XS(XS_YCP_call_ycp)
{
    // defines "items", the number of arguments
    dXSARGS; // must be called only once per xsub because it calls POPMARK

    /*
     * Plan:
     *  get the namespace name
     *  get the function name
     *  get the ns from the component broker (importing it, if necessary)
     *  find the function
     *  check its declaration, trying to convert the actual parameters
     *     to the desired ones
     *  call it
     *  pass the return value - if it's a list and perl wanted a list, so be it
     */

    // calling convention:
    //  namespace name (without trailing ::)
    //  function name
    //  function arguments
    if (items < 2 || !SvPOK (ST (0)) || !SvPOK (ST (1)))
    {
	y2internal ("Perl called YCP without specifying a namespace and a name");
	XSRETURN_EMPTY;
    }

    const char * ns_name = SvPV_nolen (ST (0));
    const char * func_name = SvPV_nolen (ST (1));

    // The perl interpreter must not get reinitialized.
    // Therefore we push what we got (aTHX) to YPerl so that it does
    // not try to construct its own.
    YPerl *yperl = YPerl::yPerl (aTHX); // access the singleton object

    YCPValue ret_yv = YCPNull ();

    // access the parameters via a vector because using dXSARGS more
    // than once per XSUB call messes up Perl call stacks
    vector<SV *> args;
    args.reserve (items - 2);
    I32 i;
    for (i = 2; i < items; ++i)
    {
	args.push_back (ST (i));
    }

    // this is a hack before the builtin namespaces get a uniform interface:
    if (! strcmp (ns_name, "SCR"))
    {
	ret_yv = YCP_call_SCR (aTHX_ func_name, args);
    }
    else
    {
	Y2Namespace *ns = getNs (ns_name, func_name);
	if (ns == NULL)
	{
	    XSRETURN_EMPTY;
	}

	// we want either a function or a variable
	// so find a symbol of an unspecified category
	TableEntry *sym_te = ns->table ()->find (func_name);

	if (sym_te == NULL)
	{
	    y2error ("No such symbol %s::%s", ns_name, func_name);
	    XSRETURN_EMPTY;
	}

	if (sym_te->sentry ()->isVariable () ||
	    sym_te->sentry ()->isReference ())
	{
	    ret_yv = YCP_getset_variable (aTHX_ ns_name, sym_te->sentry (), args);
	}
	else
	{ // no indent yet

	Y2Function *func_call = ns->createFunctionCall (func_name);
	if (func_call == NULL)
	{
	    y2error ("No such function %s::%s", ns_name, func_name);
	    XSRETURN_EMPTY;
	}

	// before we attach the parameters, we have to convert them
	// from Perl to YCP. That means knowing the destination type beforehand.
	// (Later if we get overloading, this will become messy)

	// go through the actual parameters
	unsigned j;
	for (j = 0; j < args.size (); ++j)
	{
	    // convert the value according to the expected type:

	    constTypePtr param_tp = func_call->wantedParameterType ();

	    YCPValue param_v = yperl->fromPerlScalar (args[j], param_tp);
	    if (param_v.isNull ())
	    {
		// an error has already been reported, now refine it.
		y2error ("... when passing parameter #%u to %s::%s",
			 j, ns_name, func_name);
		XSRETURN_EMPTY;
	    }

	    // Attach the parameter
	    bool ok = func_call->appendParameter (param_v);
	    if (!ok)
	    {
		// TODO really need to know the place in Perl code
		// where we were called from.
		XSRETURN_EMPTY;
	    }
	} // for each actual parameter

	bool ok = func_call->finishParameters ();
	if (!ok)
	{
	    // TODO really need to know the place in Perl code
	    // where we were called from.
	    XSRETURN_EMPTY;
	}
	// go call it now!
	y2debug ("Perl is calling %s::%s", ns_name, func_name);
	ret_yv = func_call->evaluateCall ();
	delete func_call;

	} // if not variable
    }

    if (ret_yv.isNull ())
    {
	XSRETURN_EMPTY;
    }

    y2debug ("YCP returned %s", ret_yv->toString ().c_str ());

    I32 context = GIMME_V;	// what context were we called in?
    if (context == G_VOID)
    {
	y2debug ("void context, returning nothing");
	XSRETURN_EMPTY;
    }
#if 0
    // disabled to be consistent with YPerl::call - disregard array context
    else if (context == G_ARRAY && ret_yv->isList ())
    {
	y2debug ("returning a list");
	// return a list iff they want a list and we have it
	YCPList ret_lv = ret_yv->asList ();
	int ret_len = ret_lv->size ();

	SP -= items; // reset stack pointer (declared by dXSARGS)
	EXTEND (SP, ret_len); // make room for the list
	for (int i = 0; i < ret_len; ++i)
	{
	    // put return values, no need to XPUSHs
	    PUSHs (sv_2mortal (yperl->newPerlScalar (ret_lv->value (i), false)));
	}
	XSRETURN (ret_len);
    }
#endif
    else
    {
	y2debug ("returning a scalar");
	ST (0) = sv_2mortal (yperl->newPerlScalar (ret_yv, false));
	XSRETURN (1);
    }
}


YCPValue YCP_call_SCR (pTHX_ const char * func_name, const vector<SV *>& args)
{
    // access directly the statically declared builtins
    extern StaticDeclaration static_declarations;
    string qualified_name_s = string ("SCR::") + func_name;
    const char *qualified_name = qualified_name_s.c_str ();

/*
  this does not work across namespaces
    TableEntry *bi_te = static_declarations.symbolTable ()->find (qualified_name);
    if (bi_te == NULL)
    {
	y2error ("No such builtin %s",qualified_name);
	return YCPNull ();
    }

    SymbolEntry *bi_se = bi_te->sentry ();
    assert (bi_se != NULL);
    assert (bi_se->isBuiltin ());
    declaration_t bi_dt = bi_se->declaration ();
*/
    declaration_t *bi_dt = static_declarations.findDeclaration (qualified_name);
    if (bi_dt == NULL)
    {
	y2error ("No such builtin '%s'", qualified_name);
	return YCPNull ();
    }

    // construct a builtin call using the proper overloaded builtin
    YEBuiltin *bi_call = new YEBuiltin (bi_dt);

    // attach the parameters:

    // we would like to know the destination type so that we could
    // convert eg a Perl scalar to a YCP symbol, but because the
    // builtins may be overloaded, let's say we want Any

    // maybe a special exceptional hack to make Path for the 1st argument?

    YPerl *yperl = YPerl::yPerl (aTHX); // access the singleton object
    // go through the actual parameters
    unsigned j;
    for (j = 0; j < args.size (); ++j)
    {
	// convert the value according to the expected type:
	constTypePtr param_tp = (j == 0)? Type::Path : Type::Any;

	YCPValue param_v = yperl->fromPerlScalar (args[j], param_tp);
	if (param_v.isNull ())
	{
	    // an error has already been reported, now refine it.
	    // Can't know parameter name?
	    y2error ("... when passing parameter #%u to builtin %s",
		     j, qualified_name);
	    return YCPNull ();
	}
	// Such YConsts without a specific type produce invalid
	// bytecode. (Which is OK here)
	// The actual parameter's YCode becomes owned by the function call?
	YConst *param_c = new YConst (YCode::ycConstant, param_v);
	// for attaching the parameter, must get the real type so that it matches
	constTypePtr act_param_tp = Type::vt2type (param_v->valuetype ());
	// Attach the parameter
	// Returns NULL if OK, Type::Error if excessive argument
	// Other errors (bad code, bad type) shouldn't happen
	constTypePtr err_tp = bi_call->attachParameter (param_c, act_param_tp);
	if (err_tp != NULL)
	{
	    if (err_tp->isError ())
	    {
		// TODO really need to know the place in Perl code
		// where we were called from.
		y2error ("Excessive parameter to builtin %s",
			 qualified_name);
	    }
	    else
	    {
		y2internal ("attachParameter returned %s",
			    err_tp->toString ().c_str ());
	    }
	    return YCPNull ();
	}
    } // for each actual parameter

    // now must check if we got fewer parameters than needed
    // or there was another error while resolving the overload
    constTypePtr err_tp = bi_call->finalize ();
    if (err_tp != NULL)
    {
	// apparently the error was already reported?
	y2error ("Error type %s when finalizing builtin %s",
		 err_tp->toString ().c_str (), qualified_name);
	return YCPNull ();
    }

    // go call it now!
    y2debug ("Perl is calling builtin %s", qualified_name);
    YCPValue ret_yv = bi_call->evaluate (false /* no const subexpr elim */);
    delete bi_call;

    return ret_yv;
}

/*
 * If we got no arguments, return the variable value.
 * If we got one argument, set the variable value (return YCPVoid).
 * Otherwise error (and return YCPNull)
 * ns_name is for error reporting
 */
YCPValue YCP_getset_variable (pTHX_ const char * ns_name, SymbolEntryPtr var_se, const vector<SV *>& args)
{
    YCPValue ret_yv = YCPNull ();
    unsigned n = args.size ();
    // first two args are the namespace and variable name
    if (n == 0)
    {
	// get
	ret_yv = var_se->value ();
    }
    else if (n == 1)
    {
	// set
	YPerl *yperl = YPerl::yPerl (aTHX); // access the singleton object
	YCPValue val_yv = yperl->fromPerlScalar (args[0], var_se->type ());
	if (val_yv.isNull ())
	{
	    // an error has already been reported, now refine it.
	    y2error ("... when setting value of %s::%s",
		     ns_name, var_se->name ());
	    return YCPNull ();
	}
	ret_yv = var_se->setValue (val_yv);
    }
    else
    {
	y2error ("Variable %s: don't know what to do, %u arguments",
		 var_se->name (), n);
    }
    return ret_yv;
}

/* called by XSLoader::load ('YaST::YCP') */
#ifdef __cplusplus
extern "C"
#endif
XS(boot_YaST__YCP); /* prototype to pass -Wmissing-prototypes */
XS(boot_YaST__YCP)
{
    dXSARGS;
    // get rid of warning: unused variable `I32 items'
    I32 __attribute__ ((unused)) foo = items;
    char* file = __FILE__;

    XS_VERSION_BOOTCHECK ;

    if (Y2ComponentBroker::createClient ("wfm") == 0)
    {
	y2error ("Cannot create server WFM component");
    }

    newXS("YaST::YCP::call_ycp", XS_YCP_call_ycp, file);
    newXS("YaST::YCP::y2_logger", XS_YCP_y2_logger, file);
    XSRETURN_YES;
}
