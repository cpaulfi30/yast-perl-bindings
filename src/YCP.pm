#! /usr/bin/perl -w
# Martin Vidner
# $Id$

=head1 NAME

YaST::YCP - a binary interface between Perl and YCP

=head1 SYNOPSIS

 use YaST::YCP qw(:DATA :LOGGING);

 YaST::YCP::Import ("SCR");
 my $m = SCR::Read (".sysconfig.displaymanager.DISPLAYMANAGER");
 SCR::Write (".sysconfig.kernel.CRASH_OFTEN", Boolean (1));

=head1 DATA TYPES

YaST has a richer and stricter data type system than Perl.
(TODO: see native ycp docs - what, where?)

Note that the stdio-communicating agents, based on the modules
L<YaST::SCRAgent|YaST::SCRAgent> and L<ycp|ycp>, have a similar but
not the same data type mapping.

When the language binding knows what type to expect, eg. when passing
an argument to a YCP function, it will convert a Perl scalar to the
desired type.

On the other hand, if the type is not known, expressed
in YCP as C<any>, scalars will be passed as strings. If you want
a specific data type, use one of the data classes like
L<YaST::YCP::Integer|/Integer>. Of course these work also when
the type is known.

=over 4

=item void

Has only one value, C<nil>, which is represented as C<undef>.
Any data type can have C<nil> as a value.

=item any

A union of all data types. Any data type can be assigned to it.

=item string, integer, float, boolean

B<YCP to Perl:> Becomes a scalar

B<Perl to YCP:> Any scalar will become a string
(even if it looks like a number).
Use L</String>, L</Integer>, L</Float> or L</Boolean>
if you want a specific data type.

=item list E<lt>TE<gt>

B<YCP to Perl:> A list becomes a reference to an array.

B<Perl to YCP:> A reference to an array becomes a list.
I<This was different before SL9.1 Beta1:>
Perl functions returning multiple values should not return a list
but a reference to it. YCP will always set a scalar calling context,
even if the result is assigned to a list.

=item map E<lt>T1, T2E<gt>

B<YCP to Perl:> A map becomes a reference to a hash.

B<Perl to YCP:> A reference to a hash becomes a map.

=item path

B<YCP to Perl:> NOT IMPLEMENTED YET.

B<Perl to YCP:> If a path is expected, a scalar like C<".foo.bar">
will be converted to C<.foo.bar>.
Otherwise use L</Path> (which is NOT IMPLEMENTED YET).

=item symbol

B<YCP to Perl:> Becomes a L</Symbol>.

B<Perl to YCP:> If a symbol is expected, a scalar like C<"foo">
will be converted to C<`foo>.
Otherwise use L</Symbol>.

=item term

B<YCP to Perl:> Becomes a L</Term>.

B<Perl to YCP:> Use L</Term>.

=item byteblock

B<YCP to Perl:> Becomes a scalar.

B<Perl to YCP:> If a byteblock is expected, a scalar like C<"\0\1">
will be converted to C<#[0001]>.
Otherwise use L</Byteblock>.

=item locale, block E<lt>TE<gt>, ...

Not implemented.

=back

=head1 YaST::YCP

The DATA tag (in C<use YaST::YCP qw(:DATA)>) imports the data
constructor functions such as Boolean, Symbol or Term.

=cut

package YaST::YCP;
use strict;
use warnings;
use diagnostics;

require Exporter;
our @ISA = qw(Exporter);
my @e_data = qw(Boolean Byteblock Integer Float String Symbol Term);
my @e_logging = qw(y2debug y2milestone y2warning y2error y2security y2internal);
our @EXPORT_OK = (@e_data, @e_logging, "sformat");
our %EXPORT_TAGS = ( DATA => [@e_data], LOGGING => [@e_logging] );

=head2 debug

 $olddebug = YaST::YCP::debug (1);
 YaST::YCP::...
 YaST::YCP::debug ($olddebug);

Enables miscellaneous unscpecified debugging

=cut

my $debug = 0;
sub debug (;$)
{
    my $param = shift;
    if (defined $param)
    {
	$debug = $param;
    }
    return $debug;
}


## calls boot_YaST__YCP
require XSLoader;
XSLoader::load ('YaST::YCP');

=head2 Import

 YaST::YCP::Import "Namespace";
 Namespace::foo ("bar");

=cut

sub Import ($)
{
    my $package = shift;
    print "Importing $package\n" if debug;

    no strict;
    # let it get our autoload
    *{"${package}::AUTOLOAD"} = \&YaST::YCP::Autoload::AUTOLOAD;
}

=head2 logging

These functions go via liby2util and thus use log.conf.
See also ycp::y2milestone.

 y2debug ($message)
 y2milestone ($message)
 y2warning ($message)
 y2error ($message)
 y2security ($message)
 y2internal ($message)

=cut

sub y2_logger_helper ($$)
{
    my ($level, $message) = @_;
    # look _two_ frames up
    my ($package, $filename, $line, $subroutine) = caller (2);
    # this is a XS:
    y2_logger ($level, "Perl", $filename, $line, $subroutine, $message);
}

sub y2debug ($)		{ y2_logger_helper (0, shift); }
sub y2milestone ($)	{ y2_logger_helper (1, shift); }
sub y2warning ($)	{ y2_logger_helper (2, shift); }
sub y2error ($)		{ y2_logger_helper (3, shift); }
sub y2security ($)	{ y2_logger_helper (4, shift); }
sub y2internal ($)	{ y2_logger_helper (5, shift); }

=head2 sformat

Implements the sformat YCP builtin:

C<sformat ('%2 %% %1', "a", "b")> returns C<'b % a'>

It is useful mainly for messages marked for translation.

=cut

sub sformat ($@)
{
    # don't shift
    # now the % indices can be used for @_
    my $format = $_[0];

    # g: global, replace all occurences
    # e: expression, not a string
    $format =~ s{%([1-9%])}{
	($1 eq '%') ? '%' : $_[$1]
    }ge;

    return $format;
}

# shortcuts for the data types
# for POD see packages below

sub Boolean ($)
{
    return new YaST::YCP::Boolean (@_);
}

sub Byteblock ($)
{
    return new YaST::YCP::Byteblock (@_);
}

sub Integer ($)
{
    return new YaST::YCP::Integer (@_);
}

sub Float ($)
{
    return new YaST::YCP::Float (@_);
}

sub String ($)
{
    return new YaST::YCP::String (@_);
}

sub Symbol ($)
{
    return new YaST::YCP::Symbol (@_);
}

sub Term ($@)
{
    return new YaST::YCP::Term (@_);
}

# by defining AUTOLOAD in a separate package, undefined functions in
# the main one will be detected
package YaST::YCP::Autoload;
use strict;
use warnings;
use diagnostics;

# cannot rely on UNIVERSAL::AUTOLOAD getting automatically called
# http://www.rocketaware.com/perl/perldelta/Deprecated_Inherited_C_AUTOLOAD.htm

# Gets called instead of all functions in Import'ed modules
# It assumes a normal function, not a class or instance method
sub AUTOLOAD
{
    our $AUTOLOAD;

    # strip $self on the way from Perl to YCP,
    # just as it is added in the reverse direction
    my $himself = shift;
    print "$himself $AUTOLOAD (", join (", ", @_), ")\n" if YaST::YCP::debug;

    my @components = split ("::", $AUTOLOAD);
    my $func = pop (@components);
    return YaST::YCP::call_ycp (join ("::", @components), $func, @_);
}

=head2 Boolean

 $b = YaST::YCP::Boolean (1);
 $b->value (0);
 print $b->value, "\n";
 SCR::Write (".foo", $b);

=cut

package YaST::YCP::Boolean;
use strict;
use warnings;
use diagnostics;

# a Boolean is just a blessed reference to a scalar

sub new
{
    my $class = shift;
    my $val = shift;
    return bless \$val, $class
}

# get/set
sub value
{
    # see "Constructors and Instance Methods" in perltoot
    my $self = shift;
    if (@_) { $$self = shift; }
    return $$self;
}

=head2 Byteblock

A chunk of binary data.

 use YaST::YCP qw(:DATA);

 read ($dev_random_fh, $r, 100);
 $b = Byteblock ($r);
 $b->value ("Hello\0world\0");
 print $b->value, "\n";
 return $b;

=cut

package YaST::YCP::Byteblock;
use strict;
use warnings;
use diagnostics;

# a Byteblock is just a blessed reference to a scalar
# just like Boolean, so use it!

our @ISA = qw (YaST::YCP::Boolean);

=head2 Integer

An explicitly typed integer, useful to put in heterogenous data structures.

 use YaST::YCP qw(:DATA);

 $i = Integer ("42 and more");
 $i->value ("43, actually");
 print $i->value, "\n";
 return [ $i ];

=cut

package YaST::YCP::Integer;
use strict;
use warnings;
use diagnostics;


# an Integer is just a blessed reference to a scalar
# just like Boolean, so use it!

our @ISA = qw (YaST::YCP::Boolean);

=head2 Float

An explicitly typed float, useful to put in heterogenous data structures.

 use YaST::YCP qw(:DATA);

 $f = Float ("3.41 is PI");
 $f->value ("3.14 is PI");
 print $f->value, "\n";
 return [ $f ];

=cut

package YaST::YCP::Float;
use strict;
use warnings;
use diagnostics;


# a Float is just a blessed reference to a scalar
# just like Boolean, so use it!

our @ISA = qw (YaST::YCP::Boolean);

=head2 Path

Not implemented yet.

=cut

=head2 String

An explicitly typed string, useful to put in heterogenous data structures.

 use YaST::YCP qw(:DATA);

 $s = String (42);
 $s->value (1 + 1);
 print $s->value, "\n";
 return [ $s ];

=cut

package YaST::YCP::String;
use strict;
use warnings;
use diagnostics;

# a String is just a blessed reference to a scalar
# just like Boolean, so use it!

our @ISA = qw (YaST::YCP::Boolean);

=head2 Symbol

 use YaST::YCP qw(:DATA);

 $s = Symbol ("next");
 $s->value ("back");
 print $s->value, "\n";
 return Term ("id", $s);

=cut

package YaST::YCP::Symbol;
use strict;
use warnings;
use diagnostics;


# a Symbol is just a blessed reference to a scalar
# just like Boolean, so use it!

our @ISA = qw (YaST::YCP::Boolean);

=head2 Term

 $t = new YaST::YCP::Term("CzechBox", "Accept spam", new YaST::YCP::Boolean(0));
 $t->name ("CheckBox");
 print $t->args->[0], "\n";
 UIx::OpenDialog ($t);

=cut

package YaST::YCP::Term;
use strict;
use warnings;
use diagnostics;

# a Term has a name and arguments

sub new
{
    my $class = shift;
    my $name = shift;
    my $args = [ @_ ];
    return bless { name => $name, args => $args }, $class
}

# get/set
sub name
{
    # see "Constructors and Instance Methods" in perltoot
    my $self = shift;
    if (@_) { $self->{name} = shift; }
    return $self->{name};
}

# get/set
sub args
{
    # see "Constructors and Instance Methods" in perltoot
    my $self = shift;
    if (@_) { @{ $self->{args} } = @_; }
    # HACK:
    # because I don't want to process multiple return values,
    # I return it as a reference
    return $self->{args};
}

1;
