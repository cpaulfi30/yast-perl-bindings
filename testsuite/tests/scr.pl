#! /usr/bin/perl

use warnings;
use strict;

use YaST::YCP;
YaST::YCP::Import ("SCR");

my $path = ".target.size";
my $arg = $0;
my $size = SCR::Read ($path, $arg);
print "Read ($path, $arg): $size\n";

