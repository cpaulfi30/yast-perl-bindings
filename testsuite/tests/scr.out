Read (.target.size, tests/scr.pl): 203
